### PR TITLE
fix: Improve chectl KubeHelper class and context

### PR DESCRIPTION
**Pull request by flacatus**
Signed-off-by: Flavius Lacatusu <flacatus@redhat.com>

### What does this PR do?
Improve chectl context and improve kubeHelper class to not initialize kube context in every command.

Ref issue: https://github.com/eclipse/che/issues/18684
### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by mmorhun** [comment]
Having two contexts might be very dengeros, especially with missing typings (`ctx` has `any` type)

**Comment by flacatus** [comment]
We have 2 methods in k8sCTX
 ctx[IS_OPENSHIFT] = await kube.isOpenShift()
 ctx[IS_OPENSHIFT4] = await kube.isOpenShift4()
Maybe instead of using them in ctx we can call this method without context

**Comment by mmorhun** [comment]
It is not ideal solution, but maybe we can have sort of inheritance with contexts: `initK8SCtx` includes usual chectl context.
Also, it would be nice to have some typings that defines common fields and allows arbitrary values too. For extended (K8s) conntext the fields have to be of `type | undefined`. @flacatus @tolusha WDYT?

**Comment by mmorhun** [comment]
> Maybe instead of using them in ctx we can call this method without context

Fine for me

File: src/api/kube.ts [hunk 1]
```diff
@@ -30,16 +30,7 @@ import { VersionHelper } from './version'
...
-  static initializeKubeConfig(): KubeConfig {
-    const kc = new KubeConfig()
-    kc.loadFromDefault()
-    cli.info(`› Current Kubernetes context: '${kc.currentContext}'`)
-    return kc
-  }
-
-  portForwardHelper = new PortForward(KubeHelper.KUBE_CONFIG, true)
-  logHelper = new Log(KubeHelper.KUBE_CONFIG)
+  public readonly kubeConfig
```
Comment (AndrienkoAleksandr) [inline src/api/kube.ts:33]: Maybe we should explicitly define type:

```typescript
 public readonly kubeConfig: KubeConfig
 ```

File: src/commands/server/delete.ts [hunk 2]
```diff
@@ -97,7 +98,9 @@ export default class Delete extends Command {
   }
 
   async isDeletionConfirmed(flags: any): Promise<boolean> {
-    const cluster = KubeHelper.KUBE_CONFIG.getCurrentCluster()
+    const kc = KubeHelper.initializeKubeConfig()
```
Comment (AndrienkoAleksandr) [inline src/commands/server/delete.ts:101]: Sorry, why do you need initialize kube config? I think we can get cluster directly  kc.kubeConfig.getCurrentCluster() ?

Comment (flacatus) [reply src/commands/server/delete.ts:101]:
``` javascript
const kc = new KubeHelper(flags)
const cluster = kc.kubeConfig.getCurrentCluster()
```

Comment (flacatus) [reply src/commands/server/delete.ts:101]: Is better

File: src/api/kube.ts (hunk 1 above)
Comment (flacatus) [reply src/api/kube.ts:33]: +1

File: src/commands/auth/delete.ts [hunk 3]
```diff
@@ -42,7 +42,7 @@ export default class Delete extends Command {
 
   async run() {
     const { args, flags } = this.parse(Delete)
-    await ChectlContext.init(flags, this)
+    await ChectlContext.initAndGet(flags, this)
```
Comment (mmorhun) [inline src/commands/auth/delete.ts:45]: For this and all identical changes: what is the reason for replacing `init` with `initAndGet` ?

File: src/commands/server/delete.ts [hunk 4]
```diff
@@ -51,7 +51,8 @@ export default class Delete extends Command {
 
   async run() {
     const { flags } = this.parse(Delete)
-    const ctx = await ChectlContext.initAndGet(flags, this)
+    let ctx = await ChectlContext.initAndGet(flags, this)
```
[the remark this answers is not in the document]
Comment (flacatus) [reply src/commands/server/delete.ts:54]: My mistake

File: src/commands/auth/delete.ts (hunk 3 above)
Comment (flacatus) [reply src/commands/auth/delete.ts:45]: My mistake

File: yarn.lock [hunk 5]
```diff
@@ -2350,7 +2350,7 @@ ecc-jsbn@~0.1.1:
 
 "eclipse-che-devfile-workspace-operator@git://github.com/devfile/devworkspace-operator#master":
   version "0.0.0"
-  resolved "git://github.com/devfile/devworkspace-operator#3bdccac9f3aa859d083c75a528b3d2d4556d6fdb"
+  resolved "git://github.com/devfile/devworkspace-operator#c7cbda90d959d5c73ad192ce4b627d4ebf7ed39d"
```
[the remark this answers is not in the document]
Comment (flacatus) [reply yarn.lock:2353]: No really. But always when I update deps this file change..

[the remark this answers is not in the document]
Comment (mmorhun) [reply yarn.lock:2353]: You don't have to commit it unless there is a need

File: src/commands/auth/login.ts [hunk 6]
```diff
@@ -121,25 +122,13 @@ export default class Login extends Command {
...
-      }
       loginData = { username, password }
     } else {
-      // Try to login via oc login credentials
-      // Check for oc command and oc login credentials
-      const stdout = (await execa(OPENSHIFT_CLI, ['status'], { timeout: 10000 })).stdout
-      if (stdout.startsWith('In project')) {
-        // User is logged into cluster with oc or kubectl
-        // Try to retrieve oc user token
+      const kube = new KubeHelper(flags)
```
[the remark this answers is not in the document]
Comment (flacatus) [reply src/commands/auth/login.ts:127]: Should fail if you don't provide che url

File: src/commands/auth/login.ts [hunk 7]
```diff
@@ -121,25 +121,14 @@ export default class Login extends Command {
...
+      flags.chenamespace = await findWorkingNamespace(flags)
```
Comment (tolusha) [inline src/commands/auth/login.ts:127]: Do we really need this here ?

Comment (tolusha) [reply src/commands/auth/login.ts:127]: Pls remove unnecessary invocation from all commands.

File: src/api/kube.ts [hunk 8]
```diff
@@ -51,10 +42,17 @@ export class KubeHelper {
     this.podReadyTimeout = (flags && flags.k8spodreadytimeout) ? parseInt(flags.k8spodreadytimeout, 10) : DEFAULT_K8S_POD_WAIT_TIMEOUT
     this.podDownloadImageTimeout = (flags && flags.k8spoddownloadimagetimeout) ? parseInt(flags.k8spoddownloadimagetimeout, 10) : DEFAULT_K8S_POD_WAIT_TIMEOUT
     this.podErrorRecheckTimeout = (flags && flags.spoderrorrechecktimeout) ? parseInt(flags.spoderrorrechecktimeout, 10) : DEFAULT_K8S_POD_ERROR_RECHECK_TIMEOUT
+    this.kubeConfig = KubeHelper.initializeKubeConfig()
```
Comment (tolusha) [inline src/api/kube.ts:45]:
```suggestion
    this.kubeConfig = new KubeConfig()
    this.kubeConfig.loadFromDefault()
```

File: src/tasks/platforms/api.ts [hunk 9]
```diff
@@ -25,8 +26,13 @@ export class ApiTasks {
       skip: () => flags['skip-kubernetes-health-check'],
```
Comment (tolusha) [inline src/tasks/platforms/api.ts:26]: Pls remove. You already take it into account below.

File: src/commands/dashboard/open.ts [hunk 10]
```diff
@@ -28,7 +27,6 @@ export default class Open extends Command {
 
   async run() {
     const { flags } = this.parse(Open)
-    flags.chenamespace = await findWorkingNamespace(flags)
```
Comment (tolusha) [inline src/commands/dashboard/open.ts:31]: pls. get it back.

**Review by tolusha — APPROVED** [review]

File: src/tasks/platforms/api.ts [hunk 11]
```diff
@@ -22,11 +23,15 @@ export class ApiTasks {
     let kube = new KubeHelper(flags)
     return {
       title: 'Verify Kubernetes API',
-      skip: () => flags['skip-kubernetes-health-check'],
       task: async (ctx: any, task: any) => {
         try {
-          await kube.checkKubeApi()
+          cli.info(`› Current Kubernetes context: '${await kube.currentContext()}'`)
```
Comment (mmorhun) [inline src/tasks/platforms/api.ts:28]: Not all the commands use these API tasks. Is it planned to print the message on a subset of commands only?

**Review by mmorhun — APPROVED** [review]
I added a few suggestions for improvements, other than that no more questions from my side.

File: src/tasks/platforms/api.ts (hunk 11 above)
Comment (tolusha) [reply src/tasks/platforms/api.ts:28]: Basically this task is run only when k8s api is used.

**Review by AndrienkoAleksandr — APPROVED** [review]

**Pull request merged**